tests/run/annotation_typing.pyx

```diff
@@ -34,57 +34,59 @@ def old_dict_syntax(a: list, b: "int" = 2, c: {'ctype': 'long int'} = 3, d: {'ty
     return a
 
 
-def pytypes_def(a: list, b: int = 2, c: long = 3, d: float = 4.0, n: list = None, o: Optional[tuple] = ()) -> list:
+def pytypes_def(a: list, b: int = 2, c: long = 3, d: float = 4.0, n: list = None, o: Optional[tuple] = (), v: Union[tuple, None]=()) -> list:
```

Comment:
We should also have tests here and for `cpdef` below that the `Union` and `Optional` declarations actually reject other types. Currently, we only test that they accept `tuple` and `None`, but not that they actually check the argument type at runtime. They could just do nothing.

I can see that there is a memoryview related test and another test for `float` further down, but we should try to cover some different kinds of argument types and all function types.
(The `float` test is even questionable because it unhelpfully rejects integers at runtime, but that's unrelated to this PR.)